v3/src/lib/dnd-kit/codap-dnd-context.tsx

```diff
@@ -26,6 +31,28 @@ export const CodapDndContext = ({ children }: IProps) => {
     threshold: { x: 0.05, y: 0.05 }
   }
 
+  const handleDragEnd = (e: DragEndEvent) => {
+    if (draggingDataSet.current && draggingAttributeId.current) {
+      appState.document.applyModelChange(() => {}, {
+        notify: dragEndNotification(draggingDataSet.current, draggingAttributeId.current)
+      })
+    }
+    draggingDataSet.current = null
+    draggingAttributeId.current = null
+  }
+
+  const handleDragStart = (e: DragStartEvent) => {
+    const info = getDragAttributeInfo(e.active)
+    if (info?.dataSet) {
+      const { dataSet, attributeId } = info
+      draggingDataSet.current = dataSet
+      draggingAttributeId.current = attributeId
+      appState.document.applyModelChange(() => {}, {
+        notify: dragStartNotification(dataSet, attributeId)
+      })
+    }
+  }
```

Comment:
I would move all of this into the `WebViewDropOverlay` using `useDndMonitor` so that it's all unified in one place.

Reply:
I put these in a centralized location since `dragstart` and `dragend` notifications go out to all plugins, not just a single plugin like the other notifications handled in this PR. I guess moving them into `WebViewDropOverlay` would mean that each overlay would be in charge of sending the notification to its own plugin, and each would have to keep track of the currently dragging attribute and dataset. Does that sound right?

Reply:
I was thinking they were analogous. In that case, go ahead and leave it where it is.

Reply:
Looking into this more closely, I was able to move the code into `WebViewDropOverlay` pretty easily and improve it in a number of ways, like no longer tracking the dragging attribute at all. It does feel a bit strange to have each overlay broadcast the same notification to its own plugin, but overall I think this change is much cleaner 👍